File: python-sdk/src/dispersl/handover.py
from __future__ import annotations

import json
from dataclasses import dataclass
from typing import Any


@dataclass(frozen=True)
class NextAction:
    type: str
    to_agent: str | None = None
    prompt: str | None = None


def _clean(value: Any) -> str | None:
    if not isinstance(value, str):
        return None
    v = value.strip()
    return v if v else None


def _parse_loose_object(value: Any) -> dict[str, Any]:
    if isinstance(value, dict):
        return value
    if not isinstance(value, str):
        return {}
    parsed: Any = value
    for _ in range(2):
        if not isinstance(parsed, str):
            break
        try:
            parsed = json.loads(parsed)
        except Exception:
            return {}
    return parsed if isinstance(parsed, dict) else {}


def next_action_from_tool(raw_tool: dict[str, Any]) -> NextAction:
    function_raw = raw_tool.get("function")
    function = function_raw if isinstance(function_raw, dict) else {}
    name = (
        _clean(function.get("name")) or _clean(raw_tool.get("type")) or _clean(raw_tool.get("name"))
    )
    if not name:
        return NextAction(type="none")
    if name in {"end_session", "finish_task"}:
        return NextAction(type="end")
    if name != "handover_task":
        return NextAction(type="none")

    args = _parse_loose_object(function.get("arguments", raw_tool.get("arguments")))
    merged = {**raw_tool, **args}
    to_agent = (
        _clean(merged.get("agent_name"))
        or _clean(merged.get("to_agent"))
        or _clean(merged.get("agent"))
        or _clean(merged.get("name"))
    )
    prompt = (
        _clean(merged.get("prompt"))
        or _clean(merged.get("instructions"))
        or _clean(merged.get("message"))
    )
    if not to_agent:
        return NextAction(type="none")
    return NextAction(type="handover", to_agent=to_agent, prompt=prompt)

File: typescript-sdk/src/handover.ts
export type NextAction = { type: "handover" | "end" | "none"; toAgent?: string; prompt?: string };

function clean(value: unknown): string | undefined {
  if (typeof value !== "string") return undefined;
  const trimmed = value.trim();
  return trimmed.length > 0 ? trimmed : undefined;
}

function parseLooseObject(value: unknown): Record<string, unknown> {
  if (typeof value === "object" && value !== null) return value as Record<string, unknown>;
  if (typeof value !== "string") return {};

  let parsed: unknown = value;
  for (let i = 0; i < 2; i += 1) {
    if (typeof parsed !== "string") break;
    try {
      parsed = JSON.parse(parsed);
    } catch {
      return {};
    }
  }
  return typeof parsed === "object" && parsed !== null ? (parsed as Record<string, unknown>) : {};
}

export function extractHandover(output: unknown): { toAgent?: string; prompt?: string } {
  const record = parseLooseObject(output);
  const toAgent = clean(record.agent_name) ?? clean(record.to_agent) ?? clean(record.agent) ?? clean(record.name);
  const prompt = clean(record.prompt) ?? clean(record.instructions) ?? clean(record.message);
  const result: { toAgent?: string; prompt?: string } = {};
  if (toAgent) result.toAgent = toAgent;
  if (prompt) result.prompt = prompt;
  return result;
}

export function nextActionFromTool(rawTool: unknown): NextAction {
  const record = typeof rawTool === "object" && rawTool !== null ? (rawTool as Record<string, unknown>) : {};
  const fn = typeof record.function === "object" && record.function !== null
    ? (record.function as Record<string, unknown>)
    : {};
  const name = clean(fn.name) ?? clean(record.type) ?? clean(record.name);

  if (!name) return { type: "none" };
  if (name === "end_session" || name === "finish_task") return { type: "end" };
  if (name !== "handover_task") return { type: "none" };

  const merged = { ...record, ...parseLooseObject(fn.arguments ?? record.arguments) };
  const toAgent = clean(merged.agent_name) ?? clean(merged.to_agent) ?? clean(merged.agent) ?? clean(merged.name);
  const prompt = clean(merged.prompt) ?? clean(merged.instructions) ?? clean(merged.message);
  if (!toAgent) return { type: "none" };
  return prompt ? { type: "handover", toAgent, prompt } : { type: "handover", toAgent };
}

export function mergeNextAction(current: NextAction, candidate: NextAction): NextAction {
  if (candidate.type === "none") return current;
  if (candidate.type === "end" || current.type === "end") return candidate.type === "end" ? candidate : current;
  if (candidate.type === "handover" && candidate.toAgent) return candidate;
  return current;
}
